### ISSUE TITLE
torch rpc lose device information between rpc calls

### ISSUE THREAD

**Issue by xw285cornell** [pytorch/pytorch#42544]
## 🐛 Bug

I use rpc.remote() to instantiate a trainer class on the remote worker, and then call train() function from the rref. The `__init__` function sets the cuda device, and move the model to the right device. However, the current device information is lost during the remote call of train() function. This can cause a mismatch between the model (which is on the target device) and the current master device (reset to 0). This can cause the GPU kernels to be launched at the master device (0) while the input tensors are on the target device, if the GPU kernel is not properly protected by the device guard. 

## To Reproduce

```
import torch.multiprocessing as mp
import os
import sys

import torch
import torch.distributed.rpc as rpc


class Trainer:
    def __init__(self, device_id):
        torch.cuda.set_device(device_id)
        print(f"Trainer init, device = {torch.cuda.current_device()}")

    def train(self):
        print(f"Trainer training, device = {torch.cuda.current_device()}")


def setup_worker(rank, world_size):
    rpc.init_rpc(f"worker-{rank}", rank=rank, world_size=world_size)
    rpc.shutdown()


def setup_master(rank, world_size):
    rpc.init_rpc(f"worker-{rank}", rank=rank, world_size=world_size)
    rref = rpc.remote("worker-0", Trainer, args=(2,))
    # print(rref.to_here())
    fut = rref.rpc_async().train()
    fut.wait()
    rpc.shutdown()


if __name__ == "__main__":
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "29503"

    world_size = 2
    ctx = mp.get_context("spawn")
    worker_proc = ctx.Process(target=setup_worker, args=(1, world_size))
    worker_proc.start()

    master_proc = ctx.Process(target=setup_master, args=(0, world_size))
    master_proc.start()

    master_proc.join()
    worker_proc.join()
```

1. Build on the recent pytorch master
2. run the script
Trainer init, device = 2
Trainer training, device = 0

## Expected behavior

Trainer init, device = 2
Trainer training, device = 2

## Environment

 - PyTorch Version (e.g., 1.0): recent pytorch master
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 9.2
 - GPU models and configuration:
 - Any other relevant information:

**Comment by xw285cornell** [comment]
cc. @ngimel, @mrshenli, @rohan-varma  , @pritamdamania87

**Comment by rohan-varma** [comment]
I think this might be because functions run on the remote Trainer can run in different worker threads, and `torch.cuda.set_device()` does not seem to propagate across threads (see [gist](https://gist.github.com/rohan-varma/9b9a02db6d131081209bc33b09383390)). 

As a workaround, would it suit your use case to save the device in init, and then set that device in the `train()` function? 

Also note that the RPC framework does not currently have proper GPU support, and using GPU tensors as arguments or return values isn't supported. This is something that we are planning on supporting with the TensorPipe backend.

**Comment by xw285cornell** [comment]
Thanks @rohan-varma for the quick response! It's good to know that rpc is not supporting GPU operators also. 

Yes, the workaround works for us (and we actually did this already :) ). But this is actually not very intuitively and could be hard to debug (it actually takes us a while). Just wondering if rpc can somehow save/restore the device info as part of the context.

**Comment by mrshenli** [comment]
>  But this is actually not very intuitively and could be hard to debug (it actually takes us a while).  Just wondering if rpc can somehow save/restore the device info as part of the context.

Agree. Wondering what would be a good API for this. IIUC, for the use case above, we need to preserve context across RPC requests. It might be unnecessary to always do this for all applications, so we might need an explicit API. It also might be hard to automatically detect what states should be put into the context (e.g., current device, env vars, RNG states, etc.). Does it make sense if we ask users to provide sth like a `set_context` function which will be invoked before running the RPC target function on the destination worker? For example, it could be sth like:

```python

def my_ctx_func():
    torch.cuda.set_device(1)

@rpc.context(my_ctx_func)
def create_obj():
    pass

rref = rpc.remote("work1", create_obj)

rref.rpc_async(my_ctx_func).some_func()
```

Or, alternative might be:

```python
with rpc.context(my_ctx_func):
    rref = rpc.remote("work1", create_obj)
    rref.rpc_async().some_func()
```

**Comment by xw285cornell** [comment]
Thanks @mrshenli, this makes sense to me. I personally think the alternative design seems to be more intuitive. @ngimel what do you think?

**Comment by ngimel** [comment]
Yeah, context manager seems better.
